rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/CompactView.java

```diff
@@ -166,269 +165,107 @@ private void changeSelection(String name) {
-                myNodesNs = new NodeSource(origNs);
-                myNodesNsMap.put(nsid, myNodesNs);
-            }
+        processNodes(myNodesFlow, nodes.stream().filter(node -> usedBy(node, username)).collect(Collectors.toList()));
 
-            // at this point, at least one node of this ns is being used
-
-            Iterator<String> hostsIt = myNodesNs.getHosts().keySet().iterator();
-            while (hostsIt.hasNext()) {
-                String hostid = hostsIt.next();
-                Host host = myNodesNs.getHosts().get(hostid);
-                if (!usedBy(host, username)) {
-                    hostsIt.remove();
-                    continue;
-                }
+        //         if myNodesFlow already have some nodes, but they do not became to us anymore, then delete them
+        nodes.stream()
+             .filter(node -> !usedBy(node, username))
+             .filter(node -> myNodesFlow.isNodeDrawn(node))
+             .forEach(node -> myNodesFlow.remove(node));
 
-                Iterator<String> nodesIt = host.getNodes().keySet().iterator();
-                while (nodesIt.hasNext()) {
-                    String nodeid = nodesIt.next();
-                    Node node = host.getNodes().get(nodeid);
-                    if (!usedBy(node, username)) {
-                        nodesIt.remove();
-                        continue;
-                    } else {
-                        // node used by the logged user found!
-                    }
-                }
-            }
-        }
-        return myNodesNsMap;
     }
 
-    @Override
-    public void nodesUpdated(Map<String, NodeSource> nodes) {
-        /* show only nodes used by the logged user */
-        if (onlyMyNodes) {
-            nodes = filterMyNodes(nodes);
-        }
+    private boolean usedBy(Node n, String username) {
+        return username != null && username.equals(n.getNodeOwner());
+    }
 
-        /* first call : create the components */
-        if (this.flow == null) {
-            this.flow = new FlowPanel();
-            this.flow.setWidth("100%");
-            // removes the vertical space between lines
-            this.flow.getElement().getStyle().setProperty("lineHeight", "0");
-            this.curTiles = new LinkedList<String>();
-
-            // add every tile at once, else it will attempt to render the page
-            // each time we add a new tile
-            for (NodeSource ns : nodes.values()) {
-                flow.add(new NodeTile(ns));
-                curTiles.add(ns.getSourceName());
-                for (Node n : ns.getDeploying().values()) {
-                    flow.add(new NodeTile(n));
-                    curTiles.add(n.getNodeUrl());
+    private void processNodes(CompactFlowPanel flow, List<Node> nodes) {
+        for (Node node : nodes) {
+            if (node.isRemoved()) {
+                if (flow.isNodeDrawn(node) && flow.isNodeSourceDrawn(node.getSourceName())) {
+                    flow.remove(node);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Iterate over the "entrySet" instead of the "keySet". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2864)